Add struct Path for VFS

Plucky923 opened the pull request:
This PR aims to address issue #756 . Here are the main changes:

1. **Addition of `struct Path`**:

```rust
pub struct Path {
    mount_node: Arc<MountNode>,
    dentry: Arc<Dentry>,
    this: Weak<Path>,
}
```

2. **Addition of the `parent` field to `struct MountNode`**:

```rust
pub struct MountNode {
    root_dentry: Arc<Dentry>,
    mountpoint_dentry: Option<Arc<Dentry>>,
    fs: Arc<dyn FileSystem>,
    parent: RwLock<Option<Weak<MountNode>>>,
    children: Mutex<BTreeMap<DentryKey, Arc<Self>>>,
    this: Weak<Self>,
}
```

3. **Removal of the `mount_node` field in `struct Dentry`**:

```rust
pub struct Dentry {
    inode: Arc<dyn Inode>,
    name_and_parent: RwLock<Option<(String, Arc<Dentry>)>>,
    this: Weak<Dentry>,
    children: Mutex<Children>,
    flags: AtomicU32,
}
```

4. **Movement of the pathname lookup function to `impl Path`**.
5. **Modification of some code that utilizes pathname lookup**.

Below are explanations for some notable changes:

- **`pub fn new_child`**: This function is defined in `impl Dentry`. It's utilized to create new child `struct Dentry` instances, such as in `pub fn link` (defined in `impl Path`). This encapsulation facilitates creating child `struct Dentry` instances in various contexts.
- **`pub fn lookup_fast` and `pub fn lookup_slow`**: These functions are defined in `impl Dentry` and handle pathname lookup. The pathname lookup first search for `struct Dentry` in the dentry cache (DCACHE). If not found, they proceed to search in the specific filesystem. These operations are encapsulated for ease of use.
- **`pub fn rename`**: This function is called by the `syscall rename`. A judgment within it has been moved from `pub fn rename` to the `syscall rename`. This adjustment is due to its reliance on `struct MountNode`, which is no longer directly accessible after removing the `mount_node` field from `struct Dentry`.
- **Choice of functions moved to `impl Path`**: Functions involving the `mount_node` field of `struct Dentry` have been moved to `impl Path`. Functions not directly dependent on this field, like `pub fn unlink` and `pub fn link`, have remained in their original locations to minimize code modifications.
- **`syscall rename` will call `pub fn rename` (defind in `impl Dentry`). The following judgment is moved from `pub fn rename` to `syscall rename`.**  

```rust
if !Arc::ptr_eq(&self.mount_node(), &new_dir.mount_node()) {	
     return_errno_with_message!(Errno::EXDEV, "cannot cross mount");	
}
```

   Because this judgment involves `struct MountNode`, but the `mount_node` field of `struct Dentry` is removed, it is not easy to make this judgment in `pub fn rename`. The reason why `pub fn rename` is not moved to `impl Path` is because `pub fn rename` involves a lot of Dentry operations, and moving it to `impl Path` will involve a lot of modifications.

StanPlatinum commented:
This PR has been reviewed line by line by me and @grief8 .

Plucky923 commented:
I plan to change `struct Path` to `struct VfsLocation`, and there are some problems to solve together later.

StanPlatinum commented:
> I'm considering renaming `struct Path` to `struct DentryMnt` since this structure is at the VFS layer and can uniquely determine the location in a directory tree.

I also think the name `DentryMnt` is suitable here since the type encompasses two dimensions, akin to a coordinate system where the `Dentry` is analogous to the x-axis, and the `MountNode` corresponds to the y-axis. When instantiating this type, both Dentry and MountNode are to pinpoint a specific location.

Furthermore, the methods it implements are similar to those of the previous `Dentry`, but also include some operations manipulating the `MountNode`.

tatetian commented:
@liqinggd @yingdi-shan Could you take a look at this PR? Your inputs are highly appreciated.

tatetian commented on kernel/aster-nix/src/fs/fs_resolver.rs:
```diff
@@ -37,33 +37,33 @@ impl Default for FsResolver {
 
     /// Open or create a file inode handler.
-    pub fn open(&self, path: &FsPath, flags: u32, mode: u16) -> Result<InodeHandle> {
+    pub fn open(&self, pathname: &FsPath, flags: u32, mode: u16) -> Result<InodeHandle> {
```
If `Path` is named to `DentryMnt`, then renaming `path` to `pathname` would not be necessary, right?

Plucky923 replied:
This is an issue I want to mention. In some places, the path name of the file is named path and in some places it is named pathname. In order to facilitate understanding, I think it may need to be unified, such as in
`kernel/aster-nix/src/syscall/chmod.rs` and `kernel/aster-nix/src/syscall/chdir.rs`.

tatetian commented:
It seems that `DentryMnt` would be used more often than `Dentry`, right? What are the rare situations when `Dentry` is used rather than `DentryMnt`?

Plucky923 commented:
> It seems that `DentryMnt` would be used more often than `Dentry`, right? What are the rare situations when `Dentry` is used rather than `DentryMnt`?

Plucky923 commented:
> It seems that `DentryMnt` would be used more often than `Dentry`, right? What are the rare situations when `Dentry` is used rather than `DentryMnt`?

`Dentry` is used to represent directory entries. It is mainly used to map path names to actual files or directories in the file system, including file names and inodes.
`DentryMnt` is used to locate files or directories and know their location in the file system tree and the file system mount point where they are located.
Moving many methods into `impl DentryMnt` does not mean that `Dentry` has no role. They play a role in their respective fields. For example, when we are doing pathname lookup, `DentryMnt` can better represent the current location we are looking for, while traversing the directory requires `Dentry`. This is also the reason why `lookup_fast` and `lookup_slow` are encapsulated. Of course, it also needs the help of `MountNode`. They can cooperate with each other to better complete the pathname lookup process.

StanPlatinum commented:
After a discussion with @liqinggd , we have the following comments:

1. Within the context of DentryMnt, our focus will be on operations such as lookup operations, `mount`, and `umount`. These are crucial for managing the mounting mechanism and ensuring a seamless interaction with the filesystem.

2. On the other hand, when we delve into the Dentry scope, we'll be addressing a different set of operations. This includes functionalities like `mkdir`, `rmdir`, `link`, `unlink`, and `rename`. These operations cater to the more common filesystem manipulations that do not directly involve mounting or unmounting actions.

3. Additionally, it's worth noting that if any operation requires validation or checks against a mountnode, such considerations should be appropriately handled within the syscall layer.

Plucky923 commented:
I fixed several issues commented above in the new commit.
1. The places where path names need to be expressed are named `path`.
2. Moved the `fn link` back to the `impl Dentry`. The `fn link` determined that the cross mount error requires the use of `MountNode`. Because only `sys_link` calls `fn link`, the error judgment is moved to syscall.

tatetian commented on kernel/aster-nix/src/fs/utils/dentry.rs:
```diff
@@ -147,30 +106,25 @@ impl Dentry {
     }
 
-    fn set_mountpoint(&self) {
+    pub fn set_mountpoint(&self) {
```
I suggest merging `dentrymnt.rs` into `dentry.rs`. In software engineering, we want to write modular code. The module boundaries should be set to make modules _highly cohesive_ and _loosely coupled_. This goal is not achieved by the current version. Some originally private methods of `Dentry` such as `xx_mountpoint` are made public. New helper methods like `lookup_xx` are added. But the only users of these methods are `DentryMnt`. This shows that the two types are highly coupled.

tatetian commented on kernel/aster-nix/src/fs/utils/dentrymnt.rs:
```diff
@@ -0,0 +1,191 @@
+}
+
+impl DentryMnt {
+    pub fn new(mount_node: Arc<MountNode>, dentry: Arc<Dentry>) -> Arc<Self> {
```
This constructor of `DentryMnt` is ill defined as it allows the user to instantiate `DentryMnt` with unmatched pairs of `mount_node` and `dentry`.

```rust
impl DentryMnt {
    pub fn new(mount_node: Arc<MountNode>, dentry: Arc<Dentry>) -> Arc<Self> { ... }
```

The VFS subsystem, particularly the `dentry` and `mount` modules, should and can be structured and designed so that it is impossible for the VFS users to construct invalid `DentryMnt` instances. Combining the two modules of `dentry.rs` and `dentrymnt.rs` would be a good start towards this goal.

As a reference, see how the constructors of the current `Dentry`. It does not expose a `new` method that allows constructing invalid instances of `Dentry`.

Plucky923 replied:
Thank you for your advice. Upon careful examination of all instances where `DentryMnt::new()` is utilized, I have identified two primary scenarios:

**Representing the Root of a File System:** This is necessary when indicating the root location of a file system. For instance, the initialization of the initial process's cwd and root requires representation by the root directory of the root file system. Similarly, the initialization of the devpts file system also necessitates a `DentryMnt` to denote its root location. In this scenario, `DentryMnt` serves as a tuple consisting of a file system's `MountNode` and the `root_dentry` of that `MountNode`.  
**Creating a New File and Representing Its Location:** This arises when a new file is created, and its position within the file system needs to be defined.  
To accommodate these two scenarios, we should allow users to construct meaningful instances of `DentryMnt`. To this end, I have defined two methods: `new_fs_root` for the first scenario and `new_fs_child` for the second.  
For the first case, I have defined the `new_fs_root` method to be used wherever the root directory's position of a file system needs to be represented:

```rust
pub fn new_fs_root(mount_node: Arc<MountNode>) -> Arc<Self> {  
	Self::new(mount_node.clone(), mount_node.root_dentry().clone())  
}  
```

Addressing the second case, the previous implementation involved calling the `create()` method on a `Dentry`, followed by the combination of a `MountNode` and the newly created `Dentry`. I have encapsulated this logic into the `new_fs_child` method to replace the original approach:

```rust
pub fn new_fs_child(&self, name: &str, type_: InodeType, mode: InodeMode) -> Result<Arc<Self>> {  
	let new_child_dentry = self.dentry.create(name, type_, mode)?;  
	Ok(Self::new(self.mount_node.clone(), new_child_dentry.clone()))  
}  
```

I expose these two methods to other users.

tatetian replied:
I agree with this direction of improvement.

tatetian commented on kernel/aster-nix/src/fs/utils/dentrymnt.rs:
```diff
@@ -0,0 +1,191 @@
+    }
+
+    /// Get the dentry of this dentrymnt.
+    pub fn dentry(&self) -> &Arc<Dentry> {
```
I suggest hiding `Dentry` from the users of `DentryMnt`. Everything `Dentry` can do, it can be done through `DentryMnt`.

```rust
#[inherit_methods(from = "self.dentry")]
impl DentryMnt {
    pub fn fs(&self) -> Arc<dyn FileSystem>;
    pub fn sync(&self) -> Result<()>;
    pub fn metadata(&self) -> Metadata;
    pub fn type_(&self) -> InodeType;
    pub fn mode(&self) -> Result<InodeMode>;
    pub fn set_mode(&self, mode: InodeMode) -> Result<()>;
    pub fn size(&self) -> usize;
    pub fn resize(&self, size: usize) -> Result<()>;
    pub fn owner(&self) -> Result<Uid>;
    pub fn set_owner(&self, uid: Uid) -> Result<()>;
    pub fn group(&self) -> Result<Gid>;
    pub fn set_group(&self, gid: Gid) -> Result<()>;
    pub fn atime(&self) -> Duration;
    pub fn set_atime(&self, time: Duration);
    pub fn mtime(&self) -> Duration;
    pub fn set_mtime(&self, time: Duration);
}
```

In fact, I suggest (1) combing the two dentry modules into one, (2) exposing only a `Dentry` type (currently know as `DentryMnt`), which internally relies on a private `Dentry_` or `InnerDentry` (currently know as `Dentry`).

Plucky923 replied:
Thank you very much for your guidance and suggestions. During the coding process, I noticed this issue. To implement certain methods of `DentryMnt`, I had to make methods defined in `Dentry` public, and these public methods have only one unique user, `DentryMnt`. 

Therefore, I have revisited the relationship between `DentryMnt` and `Dentry` and have merged these two modules. All the original methods in `Dentry` have been made private, with the necessary methods exposed (public) by `DentryMnt`, including methods such as `mknod`, `link`, `unlink`, and others, as well as methods inherited from `inode` by `dentry`. Regarding the naming of `DentryMnt` and `Dentry`, I believe that `DentryMnt` more clearly indicates that this structure is composed of `Dentry` and `MountNode`, as its main function is to uniquely represent a position in a directory tree using a tuple of a `Dentry` and a `MountNode`, rather than maintaining the hierarchy of the directory tree. Meanwhile, `Dentry` is primarily responsible for maintaining the hierarchy, so the name `Dentry` more clearly and intuitively represents a directory entry. 

I have also provided further explanations regarding the relationship between `DentryMnt`, `Dentry`, and `MountNode` in the latest comments.

tatetian commented on kernel/aster-nix/src/fs/utils/dentrymnt.rs:
```diff
@@ -0,0 +1,191 @@
+    /// Get the overlaid dentrymnt of self.
+    ///
+    /// It will jump into the child mount if it is a mountpoint.
+    fn overlaid_dentrymnt(&self) -> Arc<Self> {
```
I find it abrupt to see this word of "overlaid". The verb "overlay" does not appear anywhere in our current codebase or is defined by our Rust doc. The first idea that comes to my mind when I see this word is overlayfs. But apparently, the code has nothing to do with overlayfs. Could you think of a more appropriate word?

Plucky923 replied:
When multiple file systems are mounted on the same mount point, consider the following scenario: We sequentially execute the following three commands:
```shell
mount /dev/sda1 /mnt
mount /dev/sda2 /mnt
mount /dev/sda3 /mnt
```
Upon mounting `/dev/sda1` to `/mnt`, we can access the contents of `/dev/sda1` through the /mnt directory. When `/dev/sda2` is also mounted to `/mnt`, the contents of `/dev/sda1` are overshadowed (not replaced) by `/dev/sda2`. At this point, the contents accessible at `/mnt` are from `/dev/sda2`. Similarly, when `/dev/sda3` is mounted to `/mnt`, the contents of `/dev/sda2` are overshadowed by `/dev/sda3`, and the contents accessible at `/mnt` are now from `/dev/sda3`.

Unmounting the file system mounted at `/mnt` (which is `/dev/sda3` at the top) allows us to access the contents of `/dev/sda2` at `/mnt`. If we unmount the file system from `/mnt` again, we can then access the contents of `/dev/sda1`.

In this example, the file systems behave as if they are placed in a stack. The file system we can access at `/mnt` at any given time is the one at the top of the stack. The original `overlaid_dentrymnt` was used to retrieve the file system at the top. Since we always start our search from the bottom, to more clearly represent this process, I have renamed `overlaid_dentrymnt` to `get_top_dentrymnt`, which signifies obtaining the `DentryMnt` of the file system at the top of the stack.

tatetian replied:
`get_top_dentrymnt` sounds better 👍

tatetian commented on kernel/aster-nix/src/fs/utils/dentrymnt.rs:
```diff
@@ -0,0 +1,191 @@
+        }
+        if self.mount_node.parent().is_some() & self.mount_node.mountpoint_dentry().is_some() {
+            let parent_dentrymnt = Self::new(
+                self.mount_node.parent().unwrap().upgrade().unwrap().clone(),
```
Rewrite the big if-else block to the following form for improved efficiency and clarity.

```rust
let Some(parent) = self.mount_node.parent() else {
    return None;
}
let Some(mountpoint) = self.mount_node.mountpoint_dentry() else {
    return None;
}

let parent_dentrymnt = Self::new(parent.upgrade().unwrap(), mountpoint.clone())
parent_dentrymnt.effective_parent()

Plucky923 replied:
Thank you very much for your suggestion, I have followed your guidance to rewrite the existing if-else logic. It not only reduces the indentation level of the code, but also improves the readability of the code, making the logical flow more intuitive.
And I will apply this technique to my other code to further improve the code's robustness and maintainability. Thank you for your valuable time and professional advice.

tatetian commented on kernel/aster-nix/src/fs/utils/dentrymnt.rs:
```diff
@@ -0,0 +1,191 @@
+            return self.dentry.name();
+        }
+
+        if self.mount_node.parent().is_some() & self.mount_node.mountpoint_dentry().is_some() {
```
Rewrite the big if-else block to the following form for improved efficiency and clarity. See my other comment about `effective_parent`.

tatetian commented on kernel/aster-nix/src/fs/utils/dentrymnt.rs:
```diff
@@ -0,0 +1,191 @@
+        loop {
+            match dir_dentrymnt.effective_parent() {
+                None => break,
+                Some(parent_dir_dentrymnt) => {
```
To improve clarity, use the technique of _return early on error_. Do not treat the success and failure cases as equals. Handle the failure cases in the if branch, do the success cases in the main execution flow. This way, the indentation levels of code can be reduced greatly.

Compare the origin code

```rust
        loop {
            match dir_dentrymnt.effective_parent() {
                None => break,
                Some(parent_dir_dentrymnt) => {
                    path = {
                        let parent_name = parent_dir_dentrymnt.effective_name();
                        if parent_name != "/" {
                            parent_name + "/" + &path
                        } else {
                            parent_name + &path
                        }
                    };
                    dir_dentrymnt = parent_dir_dentrymnt;
                }
            }
        }
```

with the revised code

```rust
        while let Some(parent_dir_dentrymnt) = dir_dentrymnt.effective_parent() {
            path = {
                let parent_name = parent_dir_dentrymnt.effective_name();
                if parent_name != "/" {
                    parent_name + "/" + &path
                } else {
                    parent_name + &path
                }
            };
            dir_dentrymnt = parent_dir_dentrymnt;
        }
```

See how the code gets cleaner and more concise.

Plucky923 replied:
Thank you for your detailed code review and valuable suggestions. By keeping success cases within the main execution flow and simplifying the handling of errors or exceptions into loop conditions, we are able to reduce the indentation level of the code and make the main logic path clearer.
I've revised the code based on your suggestion and used a while let loop to handle success cases directly, which makes the code more concise. Your feedback helped me realize the importance of considering error handling when designing loops, and I will apply this pattern in my future coding practices.
Thanks again for your time and professional guidance, it was very helpful in improving the quality of my code.

tatetian reviewed (commented):
The code is of very high quality. I like it. I really appreciate your effort put into writing clean code.

Plucky923 commented on kernel/aster-nix/src/fs/utils/path.rs:
```diff
@@ -0,0 +1,210 @@
+
+        let path = match name {
+            "." => self.this(),
+            ".." => self.effective_parent().unwrap_or(self.this()),
```
Thank you for your review and suggestions. I have updated the code to replace the current use of `unwrap_or` with `unwrap_or_else`. Additionally, I have reviewed other parts of the code and found a similar pattern in `kernel/aster-nix/src/fs/procfs/template/dir.rs`:
``` rust
fn lookup(&self, name: &str) -> Result<Arc<dyn Inode>> {
    let inode = match name {
        "." => self.this(),
        ".." => self.parent().unwrap_or(self.this()),
```
I believe that in future modifications, we can consistently replace `unwrap_or` with `unwrap_or_else` to enhance the efficiency of our code.

Plucky923 commented on kernel/aster-nix/src/fs/utils/dentry.rs:
```diff
@@ -193,6 +147,29 @@ impl Dentry {
     }
 
+    /// Lookup a dentry from DCACHE.
+    pub fn lookup_fast(&self, name: &str) -> Option<Arc<Dentry>> {
```
Thank you for your suggestion. Renaming these two methods to `lookup_via_cache` and `lookup_via_fs` would make their purposes clearer and more immediately apparen.

Plucky923 commented:
> The code is of very high quality. I like it. I really appreciate your effort put into writing clean code. With that said, it demands some major revisions regarding code organization and API design before getting merged.

Thank you very much for your code review. Your suggestions have been very enlightening, and I've learned a lot of coding experience from them. I have checked and made modifications to all the points you mentioned. Below is my explanation and summary of the code changes.

# Necessity of this PR

Firstly, I would like to provide further clarification on why the addition of `struct DentryMnt` is required. In the previous code, there was a one-to-one correspondence between Dentry and MountNode. However, the system can encounter situations where a single `Dentry` is associated with multiple `MountNode`. A `Dentry` represents a directory entry, while a `MountNode` represents a single mount operation. Here are three scenarios where this one-to-many relationship can occur:

**Mounting a File System at Multiple Points:** For example, the following commands can be executed:

```shell
mount /dev/sda1 /mnt1  
mount /dev/sda1 /mnt2  
```

This allows access to the contents of `/dev/sda1` from both `/mnt1` and `/mnt2`. As a result, two separate `MountNode` are created from the two mount operations, leading to a situation where one `Dentry` corresponds to multiple MountNodes.

**Mount Namespaces:** Each process has its own mount tree. When a new mount namespace is created, there is a cloning operation for `MountNode`, but `Dentry` are not cloned. This can also result in a single `Dentry` being associated with multiple `MountNode`.

**Bind Mounts:** Bind mounts can also lead to the cloning of `MountNode`. This operation involves making a directory mount itself into another location, which can create additional `MountNode` that reference the same `Dentry`.

In the previous implementation, a single `Dentry` object was used to denote a location within a directory tree. `Dentry` also played a pivotal role during the `path lookup` process, representing the position of intermediate results and facilitating the search for subsequent locations. Moreover, `Dentry` was responsible for maintaining the hierarchical relationships between different directory entries.  
However, with the three scenarios outlined above, it became evident that relying on a single `Dentry` to uniquely identify a location was insufficient. For example, in the first scenario, it would be ambiguous to determine whether the content of the file system was being accessed through `/mnt1` or `/mnt2`. This ambiguity highlighted the need for a more precise way to represent a location within the directory structure.  

Consequently, the decision was made to introduce a new structure, `DentryMnt`, which is essentially a pair consisting of a `Dentry` and a `MountNode`. This composite structure allows for a more accurate representation of a position within a directory tree, addressing the complexities and ambiguities that arise from multiple mount points, mount namespaces, and bind mounts. The `DentryMnt` ensures that each directory position can be uniquely identified, taking into account the specific mount context, thereby enhancing the integrity and clarity of the file system's directory tree representation.

# The relationship between Dentry, DentryMnt and MountNode

Building upon the previous explanation of the functionality of `DentryMnt`, it is designed to specifically represent a concrete location within a directory tree. The process of `path lookup` demands high precision in determining positions, hence the relevant methods for path lookup are also defined within `DentryMnt`.

`Dentry` is focused on maintaining the hierarchical structure among directories. The methods `lookup_via_cache` and `lookup_via_fs` are utilized during `path lookup` to find the child `Dentry` of the current position's `Dentry`. Since the `Dentry` involved in this process belongs to the same `MountNode`, there is no need for `MountNode` to be involved. These methods are crucial in the `path lookup` process and are therefore defined within `Dentry`.

Given that the sole user of the methods within `Dentry` is `DentryMnt`, it indicates a high degree of coupling between `Dentry` and `DentryMnt`. If other users wish to utilize methods related to `Dentry` (such as `mknod`, `link`, `unlink`, etc., as well as methods inherited from inode by `Dentry`), they can do so through `DentryMnt`. The original methods within `Dentry` have been modified to be private, with the methods in `DentryMnt` being exposed to other users.

`MountNode` represents a single mount operation and maintains the hierarchy of mounts. A `MountNode` is created whenever a file system is mounted, and there is a close relationship between `MountNode` and `Dentry`. Each MountNode has a `root_dentry` field that represents the root directory of the mounted file system. `Dentry` maintains the hierarchical relationships between directories, and `Dentry` from different file systems do not connect directly. Instead, they are indirectly linked through `MountNode`. Once a mount is completed, the `Dentry` of two different file systems are connected via `MountNode`, allowing access to the content of the mounted file system.

Plucky923 commented:
I would like to provide further clarification regarding the modifications made to `MountNode`. In the previous code, `MountNode` was not organized as a tree structure, which made the `path lookup` process heavily reliant on the `mount_node` field of `Dentry` and the `mountpoint_dentry` field within `MountNode`. To transition from a one-to-one correspondence between `Dentry` and `MountNode` to a one-to-many relationship, we have removed the `mount_node` field from `Dentry`. Consequently, I have structured `MountNode` into a tree. This approach serves two purposes: it more clearly represents the hierarchy of mounts, and it also addresses the issues that arose from the removal of the `mount_node` field from `Dentry`. Looking ahead, we plan to gradually introduce features such as `bind mount` and `move mount`, which will involve modifications to the existing mount hierarchy. For instance, `move mount` would entail moving a mounted file system to another mount point, signifying that the `parent` and `mountpoint_dentry` of an already created `MountNode` might need to be altered. The current codebase requires a `Dentry` to instantiate a `MountNode`, and the `parent` and `mountpoint_dentry` fields of `MountNode` are immutable. To facilitate the addition of future functionalities and to enhance the flexibility of the code, these aspects have also been optimized.

In the latest code modifications, @StanPlatinum  also played an irreplaceable role, offering many suggestions for my code changes and reviewing the modified code.

tatetian commented:
@Plucky923 Thanks for the explanation and clarification! They sound great to me. I will review the revised version.

tatetian commented on kernel/aster-nix/src/fs/utils/dentry.rs:
```diff
@@ -484,13 +349,13 @@ struct Children {
 
 impl Children {
-    pub fn new() -> Self {
+    fn new() -> Self {
```
Please keep the public methods of private types such as `Children::new` as public. Yes, the Rust language allows two types under the same module to see each other's private fields and to use each other's private methods. So a private type in a module such as `Children` cannot really hide its private fields and methods from other types in the same module. But still it is a good practice to distinguish between public and private methods of a type by annotating the public ones with `pub` properly. This is because it conveys the semantics-level visibility (instead of language-level) to the developers and thereby improve code readability and maintainability.

Examine all changes that remove `pub`.

Plucky923 replied:
Thank you for your code review. I hadn't considered this issue before. To improve code readability and maintainability, I reviewed all changes that removed pub. For the private types in the `path` module, I have restored the `pub` annotations for their methods to their previous state. For methods within the `Dentry_` struct, I kept methods private if they are only used within the `Dentry_` struct itself. For methods that are intended to be used by other users, such as the `Dentry` struct, I made them `pub`. I applied the same principle to the methods in the `Dentry` struct, ensuring that they are appropriately annotated based on their intended usage.

tatetian commented on kernel/aster-nix/src/fs/utils/dentry.rs:
```diff
@@ -549,3 +411,256 @@ fn write_lock_children_on_two_dentries<'a>(
+
+/// The DentryMnt can represent a location in the mount tree.
+#[derive(Debug)]
+pub struct DentryMnt {
```
I suggest extracting `mount.rs` and `dentry.rs` into a separate module. Candidate names for the module include `path` or `tree`.

```
fs/
    path/
        mount.rs
        dentry.rs
        mod.rs
```

The rationale is that the combination of mount points and dentries enables one to look up the path of an inode in a FS or, put it another way, form the in-memory tree of the entire file system. I examined all current usage of `Dentry`: it is only used by `dentry.rs` itself and `mount.rs`. After combining `mount.rs` and `dentry.rs` under one module, the concept of `Dentry` can be hidden under the module. Outsiders only need to know `DentryMnt`.

Now that the rest of the VFS and kernel only knows `DentryMnt` (not `Dentry`), it becomes obvious that `DentryMnt` should be named `Dentry` and the original `Dentry` named `Dentry_` or `DentryInner`. This information hiding technique reduces the complexity of the entire fs subsytem.

Plucky923 replied:
Thank you for your suggestion. I noticed your earlier recommendation and have given it further consideration. Initially, I retained the name `DentryMnt` to clearly indicate that this structure comprises both `Dentry` and `MountNode`. However, after reevaluating your suggestion, I recognized its merit.

`DentryMnt`, `Dentry`, and `MountNode` are three closely related structures. The methods of `Dentry` are provided to other users via `DentryMnt`. In previous versions, significant modifications involved the `DentryMnt` structure, and future contributions might conflict with these changes. Your suggestion greatly reduces the amount of code modification and ensures future contributors are not confused by these changes. I have learned a lot from your advice and have made the corresponding changes: `DentryMnt` has been renamed to `Dentry`, and `Dentry` to `Dentry_`.

Regarding the use of `Dentry` and `Dentry_` in `MountNode`, I have thought this through as well. For example, the `mountpoint_dentry` and `root_dentry` fields in `MountNode` clearly indicate a `Dentry_` by using the `_dentry` suffix. In contrast, fields like `mountpoint` in `MountNode` represent a `Dentry`. This distinction is maintained consistently throughout the code. In `dentry.rs`, I applied the same principle: methods like `set_mountpoint` operate on `Dentry`, while `set_mountpoint_dentry` operates on `Dentry_`. I believe this naming convention clearly conveys the purpose of each method.

Again, thank you for your valuable feedback.

tatetian reviewed (commented):
Thanks for revising the PR. I think we are really close to merging it. I have only raised two minor issues.

tatetian commented on kernel/aster-nix/src/fs/path/mod.rs:
```diff
@@ -0,0 +1,6 @@
+// SPDX-License-Identifier: MPL-2.0
+pub use dentry::{Dentry, DentryKey, Dentry_};
```
We don't want to re-export `Dentry_`, right?

Plucky923 replied:
Thank you for your comment. You’re right, we don't need to re-export `Dentry_`. I will update the code to only export `Dentry` and `DentryKey`.

tatetian commented on kernel/aster-nix/src/fs/path/mod.rs:
```diff
@@ -0,0 +1,6 @@
+// SPDX-License-Identifier: MPL-2.0
```
The license header should be followed with a blank line.

Plucky923 replied:
Thank you for pointing that out. I will add a blank line after the license header.Additionally, I noticed the same issue in some other files, such as `/kernel/aster-nix/src/fs/mod.rs`. I think it would be a good idea to review these files.

tatetian reviewed (commented):
Just two minor issues.

tatetian commented on kernel/aster-nix/src/fs/path/mod.rs:
```diff
@@ -0,0 +1,6 @@
+// SPDX-License-Identifier: MPL-2.0
```
Having module-level Rust documentation would be good.

```
//! Form file paths within and across FSes with dentries and mount points.
```

Plucky923 replied:
Thank you for the suggestion! I've added module-level documentation to explain this module.

tatetian reviewed (approved):
LGTM. Thanks for the contribution!
